File: digital-marketing/backend/app/routers/payments.py
from datetime import datetime, timedelta

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app import models, schemas
from app.auth import require_role
from app.database import get_db

router = APIRouter(prefix="/api", tags=["payments"])


@router.post("/payments", status_code=201)
def record_payment(
    req: schemas.RecordPaymentRequest,
    current_user: models.User = Depends(require_role("brand", "influencer", "admin")),
    db: Session = Depends(get_db),
):
    payment = models.Payment(
        type=req.type,
        user_id=current_user.id,
        ref_id=req.ref_id or 0,
        gross=req.gross,
        amount=req.amount,
        note=req.note or "",
    )
    db.add(payment)
    db.commit()
    db.refresh(payment)
    return payment


@router.post("/subscriptions")
def subscribe(
    req: schemas.SubscribeRequest,
    current_user: models.User = Depends(require_role("influencer")),
    db: Session = Depends(get_db),
):
    profile = db.query(models.InfluencerProfile).filter(
        models.InfluencerProfile.user_id == current_user.id
    ).first()
    if profile:
        profile.plan = "pro"
        profile.plan_type = req.plan_type or "subscription"
        profile.subscribed_until = datetime.utcnow() + timedelta(days=30)

    db.add(models.Payment(
        type="influencer_subscription",
        user_id=current_user.id,
        gross=1499,
        amount=1499,
        note="Creator Pro monthly subscription",
    ))
    db.commit()
    return {"status": "subscribed", "plan": "pro"}

File: digital-marketing-RUST/backend/src/handlers/payments.rs
use axum::{
    extract::State,
    http::{HeaderMap, StatusCode},
    response::IntoResponse,
    Json,
};
use chrono::{Duration, Utc};
use serde::Deserialize;
use serde_json::json;

use crate::{
    auth::extract_user,
    errors::{AppError, AppResult},
    models::Payment,
    AppState,
};

const PAY_COLS: &str =
    "id, type AS payment_type, user_id, ref_id, \
     CAST(gross AS FLOAT8) AS gross, CAST(amount AS FLOAT8) AS amount, \
     status, note, created_at";

#[derive(Deserialize)]
pub struct RecordPaymentRequest {
    pub r#type: String,
    pub ref_id: Option<i64>,
    pub gross: f64,
    pub amount: f64,
    pub note: Option<String>,
}

pub async fn record_payment(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(req): Json<RecordPaymentRequest>,
) -> AppResult<impl IntoResponse> {
    let user = extract_user(&headers, &state.config.jwt_secret, &state.db).await?;
    if user.role == "admin" || user.role == "brand" || user.role == "influencer" {
        // allowed
    } else {
        return Err(AppError::forbidden("authenticated users only"));
    }

    let payment: Payment = sqlx::query_as::<_, Payment>(&format!(
        "INSERT INTO payments (type, user_id, ref_id, gross, amount, note) \
         VALUES ($1,$2,$3,$4,$5,$6) RETURNING {}",
        PAY_COLS
    ))
    .bind(&req.r#type)
    .bind(user.id)
    .bind(req.ref_id.unwrap_or(0))
    .bind(req.gross)
    .bind(req.amount)
    .bind(req.note.as_deref().unwrap_or(""))
    .fetch_one(&state.db)
    .await?;

    Ok((StatusCode::CREATED, Json(json!(payment))))
}

#[derive(Deserialize)]
pub struct SubscribeRequest {
    pub plan_type: Option<String>,
}

pub async fn subscribe(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(req): Json<SubscribeRequest>,
) -> AppResult<impl IntoResponse> {
    let user = extract_user(&headers, &state.config.jwt_secret, &state.db).await?;
    if user.role != "influencer" {
        return Err(AppError::forbidden("influencers only"));
    }

    let subscribed_until = Utc::now() + Duration::days(30);
    sqlx::query(
        "UPDATE influencer_profiles \
         SET plan = 'pro', plan_type = $1, subscribed_until = $2 WHERE user_id = $3",
    )
    .bind(req.plan_type.as_deref().unwrap_or("subscription"))
    .bind(subscribed_until)
    .bind(user.id)
    .execute(&state.db)
    .await?;

    sqlx::query(
        "INSERT INTO payments (type, user_id, gross, amount, note) \
         VALUES ('influencer_subscription', $1, 1499, 1499, 'Creator Pro monthly subscription')",
    )
    .bind(user.id)
    .execute(&state.db)
    .await?;

    Ok(Json(json!({ "status": "subscribed", "plan": "pro" })))
}
